src/tile.js

```diff
@@ -327,24 +426,283 @@ $.Tile.prototype = {
+     * Cache key for main cache that is 'cache-equal', but different from original cache key
+     * @return {string}
+     * @private
+     */
+    buildDistinctMainCacheKey: function () {
+        return this.cacheKey === this.originalCacheKey ? "mod://" + this.originalCacheKey : this.cacheKey;
+    },
+
+    /**
+     * Read tile cache data object (CacheRecord)
+     * @param {string} [key=this.cacheKey] cache key to read that belongs to this tile
+     * @return {OpenSeadragon.CacheRecord}
+     */
+    getCache: function(key = this._cKey) {
+        const cache = this._caches[key];
+        if (cache) {
+            cache.withTileReference(this);
+        }
+        return cache;
+    },
+
+    /**
+     * Create tile cache for given data object. NOTE: if the existing cache already exists,
+     * data parameter is ignored and inherited from the existing cache object.
+     *
+     * @param {string} key cache key, if unique, new cache object is created, else existing cache attached
+     * @param {*} data this data will be IGNORED if cache already exists; therefore if
+     *   `typeof data === 'function'` holds (both async and normal functions), the data is called to obtain
+     *   the data item: this is an optimization to load data only when necessary.
+     * @param {string} [type=undefined] data type, will be guessed if not provided (not recommended),
+     *   if data is a callback the type is a mandatory field, not setting it results in undefined behaviour
+     * @param {boolean} [setAsMain=false] if true, the key will be set as the tile.cacheKey,
+     *   no effect if key === this.cacheKey
+     * @param [_safely=true] private
+     * @returns {OpenSeadragon.CacheRecord|null} - The cache record the tile was attached to.
+     */
+    addCache: function(key, data, type = undefined, setAsMain = false, _safely = true) {
+        const tiledImage = this.tiledImage;
+        if (!tiledImage) {
+            return null; //async can access outside its lifetime
+        }
+
+        if (!type) {
+            if (!this.__typeWarningReported) {
+                $.console.warn(this, "[Tile.addCache] called without type specification. " +
+                    "Automated deduction is potentially unsafe: prefer specification of data type explicitly.");
+                this.__typeWarningReported = true;
+            }
+            if (typeof data === 'function') {
+                $.console.error("[TileCache.cacheTile] options.data as a callback requires type argument! Current is " + type);
+            }
+            type = $.convertor.guessType(data);
+        }
+
+        const overwritesMainCache = key === this.cacheKey;
+        if (_safely && (overwritesMainCache || setAsMain)) {
+            // Need to get the supported type for rendering out of the active drawer.
+            const supportedTypes = tiledImage.viewer.drawer.getSupportedDataFormats();
+            const conversion = $.convertor.getConversionPath(type, supportedTypes);
```

Comment:
I think this likely to be buggy right now, because `WebGLDrawer` uses a `CanvasDrawer` fallback for certain cases, like when cross-origin data doesn't allow `webgl` to be used. And this can happen for only some `TiledImage`s within a given viewer; so, you'd need to know specifically what type of format the drawer needs not just in general but for a specific tiled image, and this can change dynamically after a CORS error is detected.

I have an idea I think would facilitate this process by moving the logic away from here and onto the drawer, which is where I think it more naturally belongs. I'll lay that idea out elsewhere.

Reply:
Nice point. If you can do it please do so. Hovever, note that third party code must be able to interact flawlessly. If system decides that it has to free data, the drawer must allow this. I guess with flexible enough interface this is doable.

Reply:
Also note that the current approach accounts for less crazy drawers, that just say 'I support rendering using context2d object'. And everything is handled for them. Only webgl drawer needs a special conversion step, which does not even hold the data on CPU. This must be allowed since we allow incremental updates: you can opt for starting invalidation pipeline from the current viewport data. This means you have to be able to convert texture to e.g. context2d / image. This is done by using _internal cache_ that holds logics of loading textures, and externally the record is just a reference to the data that was uploaded as a texture. i.e. using pseudonotation:
 `CacheRecord{ type: 'contex2d_TEXTURE_2D', data: context2d, internalCache: SimpleCacheRecord{ type: 'TEXTURE_2D'} }`

Reply:
One more reason why this will not work is that  `cacheRecord.addHandler('cache-data-unloaded', (event)=> cleanup(event.target._webglTexture) )`  is not necessarily called every time you need it. The tile cache can decide that we need to free the texture, not the data itself. E.g. if original cache is TEXTURE_2D_CONTEXT2D (internal T2D, outer c2d), it frees the internal cache and conversion to context2d type is done. If you want to on the other hand convert CONTEXT2D original cache to a CONTEXT2D, it will stimply do nothing since it is already done. But there was a private property `_webglTexture `, now being a memory leak (because conversion trigger meant new data will be used for rendering instead, and this cache will never get used again). But as said, if we used internal cache for storing the tex reference, it is not a problem, except we need to change the cache type too (or force-wipe internal cache on any conversion, even when there is nothing to convert, which feels ugly design to me RN as I am writhing this).

Reply:
> Event on the data removal means possible collision (might happen multiple parties think its _their_ job to free the data)

An actual implementation of this could solve this problem in various ways, including by passing in/attaching a destructor like in your current implementation.

> and leads to unclear code (attaching custom properties to cache, e.g. `cacheRecord._webglTexture = tex;`, I've seen these custom references on the `Tile` instance all over the place, and wanted to avoid these).

I agree, and I'm not advocating for attaching arbitrary custom properties. It can be done with other ways such as an official API on CacheRecord. The point is just that a conversion initiated by the drawer at the time of drawing can be cached and managed by the cache.

The problem we need to address - which is not yet solved - is that multiple drawers can be used at the same time within a single viewer (at the moment, this is primarily because some `TiledImage`s cannot be supported by `webgl`). This means that `Tile`s belonging to different `TiledImage`s need to be able to have different data types. Having the cache system require a single data type doesn't support this. To me it makes the most sense for the drawer to initiate the conversion (synchronously) as needed. It should be synchronous because at the time of drawing, `TiledImage` has already decided which tiles should be drawn based on the presence of the data, so async would cause potential visual bugs.

Reply:
>  I'm not advocating for attaching arbitrary custom properties

I understand this is a pseudocode, but despite being so this shows what would developers do when presented with events. That is my argument: it leads to doing exactly this.

>This means that Tiles belonging to different TiledImages need to be able to have different data types.

The same data involves all tiled images the tile appears in, and this transitively makes all tiled images also tainted. No need to have different types on the tile. It is a [equivalence relation](https://en.wikipedia.org/wiki/Equivalence_relation). If this is not the case, then there is a bug. If the reasoning is that 'previously, the data was uploaded as a texture, but suddenly a tained bit of data came and now we need to swap suddenly to no-webgl', then it is not supported correctly in the current master either. If 100th tile suddenly becomes tainted, involving the whole tiled image, who frees the 99 from gpu? I have feeling that this should get also support from the internal logics (following the references and re-converting the tiles to different format that drawer defines for case it gets a tained data, for example). If you don't care about data freeing, what prevents you from grabbing the reference to the data that was moved to a texture instead?

An if this cannot happen, and all tiles of a tiled image are either tained or not, then there is no problem, it can detect this inside any converter and adjust accordingly.

Am I missing something?

Reply:
> But other TiledImages with non-tainted data would still need to be converted for the WebGLDrawer

Yes. I forgot about the case where tainted tile A belongs both to T1 and T2, B to T2 and T3, so tainted are T1 and T2 only. You are right. But it's still doable. Solving the particular case of a Webgl drawer, it can just check whether tile is tained and otherwise upload it onto GPU. If it keeps reference to the data it used to upload as well as the potential texture id, it can choose which element to use from the cache entry. 

In general, drawers can provide any kind of a type (e.g. a timestamp) to ensure it won't be used from the outside, and set 'internalCache' true to hide this cache from users. Such cache can keep references to the data in the outer cache, and just do nothing when the destructor is called. And if the drawer decides to perform it's own conversions synchronously before data is being drawn, we cannot stop them - though there is no API for freeing such data if necessary in that case.

Reply:
I added code that implements this on the current WebGL drawer.